README.md
# docker-machine-driver-metal

[![GitHub release](https://img.shields.io/github/release/equinix/docker-machine-driver-metal/all.svg?style=flat-square)](https://github.com/equinix/docker-machine-driver-metal/releases)
[![Go Report Card](https://goreportcard.com/badge/github.com/equinix/docker-machine-driver-metal)](https://goreportcard.com/report/github.com/equinix/docker-machine-driver-metal)
[![Slack](https://slack.equinixmetal.com/badge.svg)](https://slack.equinixmetal.com)
[![Twitter Follow](https://img.shields.io/twitter/follow/equinixmetal.svg?style=social&label=Follow)](https://twitter.com/intent/follow?screen_name=equinixmetal)
![](https://img.shields.io/badge/Stability-Maintained-green.svg)

The [Equinix Metal](https://metal.equinix.com) cloud bare-metal machine driver for Docker.

This repository is [Maintained](https://github.com/packethost/standards/blob/main/maintained-statement.md) meaning that this software is supported by Equinix Metal and its community - available to use in production environments.

## Usage

You can provision bare-metal hosts once you have built and installed the docker-machine driver. The binary will be placed in your `$PATH` directory.

Test that the installation worked by typing in:

```sh
docker-machine create --driver metal
```

### Example usage

This creates the following:

* Type0 machine
* in the EWR region (NJ)
* with Ubuntu 16.04
* in project $PROJECT
* Using $API_KEY - [get yours from the Portal](https://console.equinix.com/users/me/api-keys)

```sh
$ docker-machine create sloth \
  --driver metal --metal-api-key=$API_KEY --metal-os=ubuntu_16_04 --metal-project-id=$PROJECT --metal-facility-code "ewr1" --metal-plan "baremetal_0"
  
Creating CA: /home/alex/.docker/machine/certs/ca.pem
Creating client certificate: /home/alex/.docker/machine/certs/cert.pem
Running pre-create checks...
Creating machine...
(sloth) Creating SSH key...
(sloth) Provisioning Equinix Metal server...
(sloth) Created device ID $PROJECT, IP address 147.x.x.x
(sloth) Waiting for Provisioning...
Waiting for machine to be running, this may take a few minutes...
Detecting operating system of created instance...
Waiting for SSH to be available...
Detecting the provisioner...
Provisioning with ubuntu(systemd)...
Installing Docker...
Copying certs to the local machine directory...
Copying certs to the remote machine...
Setting Docker configuration on the remote daemon...
Checking connection to Docker...
Docker is up and running!
To see how to connect your Docker Client to the Docker Engine running on this virtual machine, run: docker-machine env sloth
```

> Provision time can take several minutes

At this point you can now `docker-machine env sloth` and then start using your Docker bare-metal host!

## Development

### Building

Pre-reqs: `docker-machine` and `make`

* Install the Golang SDK [https://golang.org/dl/](https://golang.org/dl/) (at least 1.11 required for [modules](https://github.com/golang/go/wiki/Modules) support

* Download the source-code with `git clone http://github.com/equinix/docker-machine-driver-metal.git`

* Build and install the driver:

```sh
cd docker-machine-driver-metal
make
sudo make install
```

Now you will now be able to specify a `-driver` of `metal` to `docker-machine` commands.

### Debugging

To monitor the Docker debugging details and the Equinix Metal API calls:

```sh
go build
PACKNGO_DEBUG=1 PATH=`pwd`:$PATH docker-machine \
  --debug create -d metal \
  --metal-api-key=$METAL_AUTH_TOKEN \
  --metal-project-id=$METAL_PROJECT \
  foo
```

### Release Process

This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

Releases are handled by [GitHub Workflows](.github/workflows/release.yml) and [goreleaser](.goreleaser.yml).

To push a new release, checkout the commit that you want released and: `make tag version=v0.2.3`.  Robots handle the rest.

Maintainers should verify that the release notes convey to users all of the notable changes between releases, in a human readable way.
The format for each release should be based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).

## Releases and Changes

See <https://github.com/equinix/docker-machine-driver-metal/releases> for the latest releases, install archives, and the project changelog.


CHANGELOG.md
See <https://github.com/equinix/docker-machine-driver-metal/releases>


pkg/drivers/metal/metal_test.go
package metal

import (
	"testing"

	"github.com/docker/machine/libmachine/drivers"
	"github.com/stretchr/testify/assert"
)

func TestSetConfigFromFlags(t *testing.T) {
	driver := NewDriver("", "")

	checkFlags := &drivers.CheckDriverOptions{
		FlagsValues: map[string]interface{}{
			"metal-api-key":    "APIKEY",
			"metal-project-id": "PROJECT",
		},
		CreateFlags: driver.GetCreateFlags(),
	}

	err := driver.SetConfigFromFlags(checkFlags)

	assert.NoError(t, err)
	assert.Empty(t, checkFlags.InvalidFlags)
}


pkg/drivers/packet/packet_test.go
package packet

import (
	"testing"

	"github.com/docker/machine/libmachine/drivers"
	"github.com/stretchr/testify/assert"
)

func TestSetConfigFromFlags(t *testing.T) {
	driver := NewDriver("", "")

	checkFlags := &drivers.CheckDriverOptions{
		FlagsValues: map[string]interface{}{
			"packet-api-key":    "APIKEY",
			"packet-project-id": "PROJECT",
		},
		CreateFlags: driver.GetCreateFlags(),
	}

	err := driver.SetConfigFromFlags(checkFlags)

	assert.NoError(t, err)
	assert.Empty(t, checkFlags.InvalidFlags)
}


driver.go
// SPDX-License-Identifier: BSD-3-Clause

package packet

import (
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/docker/machine/libmachine/drivers"
	"github.com/docker/machine/libmachine/log"
	"github.com/docker/machine/libmachine/mcnflag"
	"github.com/docker/machine/libmachine/ssh"
	"github.com/docker/machine/libmachine/state"
	"github.com/packethost/packngo"
)

const (
	dockerConfigDir = "/etc/docker"
	consumerToken   = "24e70949af5ecd17fe8e867b335fc88e7de8bd4ad617c0403d8769a376ddea72"
)

var (
	// version is set by goreleaser at build time
	version = "devel"

	// build time check that the Driver type implements the Driver interface
	_ drivers.Driver = &Driver{}
)

type Driver struct {
	*drivers.BaseDriver
	ApiKey                  string
	ProjectID               string
	Plan                    string
	HardwareReserverationID string
	Facility                string
	OperatingSystem         string
	BillingCycle            string
	DeviceID                string
	UserData                string
	Tags                    []string
	CaCertPath              string
	SSHKeyID                string
	UserDataFile            string
	UserAgentPrefix         string
	SpotInstance            bool
	SpotPriceMax            float64
	TerminationTime         *packngo.Timestamp
}

// NewDriver is a backward compatible Driver factory method.  Using
// new(packet.Driver) is preferred.
func NewDriver(hostName, storePath string) *Driver {
	return &Driver{
		BaseDriver: &drivers.BaseDriver{
			MachineName: hostName,
			StorePath:   storePath,
		},
	}
}

func (d *Driver) GetCreateFlags() []mcnflag.Flag {
	return []mcnflag.Flag{
		mcnflag.StringFlag{
			Name:   "packet-api-key",
			Usage:  "Packet api key",
			EnvVar: "PACKET_API_KEY",
		},
		mcnflag.StringFlag{
			Name:   "packet-project-id",
			Usage:  "Packet Project Id",
			EnvVar: "PACKET_PROJECT_ID",
		},
		mcnflag.StringFlag{
			Name:   "packet-os",
			Usage:  "Packet OS",
			Value:  "ubuntu_16_04",
			EnvVar: "PACKET_OS",
		},
		mcnflag.StringFlag{
			Name:   "packet-facility-code",
			Usage:  "Packet facility code",
			Value:  "ewr1",
			EnvVar: "PACKET_FACILITY_CODE",
		},
		mcnflag.StringFlag{
			Name:   "packet-plan",
			Usage:  "Packet Server Plan",
			Value:  "baremetal_0",
			EnvVar: "PACKET_PLAN",
		},
		mcnflag.StringFlag{
			Name:   "packet-hw-reservation-id",
			Usage:  "Packet Reserved hardware ID",
			EnvVar: "PACKET_HW_ID",
		},
		mcnflag.StringFlag{
			Name:   "packet-billing-cycle",
			Usage:  "Packet billing cycle, hourly or monthly",
			Value:  "hourly",
			EnvVar: "PACKET_BILLING_CYCLE",
		},
		mcnflag.StringFlag{
			Name:   "packet-userdata",
			Usage:  "Path to file with cloud-init user-data",
			EnvVar: "PACKET_USERDATA",
		},
		mcnflag.BoolFlag{
			Name:   "packet-spot-instance",
			Usage:  "Request a Packet Spot Instance",
			EnvVar: "PACKET_SPOT_INSTANCE",
		},
		mcnflag.StringFlag{
			Name:   "packet-spot-price-max",
			Usage:  "The maximum Packet Spot Price",
			EnvVar: "PACKET_SPOT_PRICE_MAX",
		},
		mcnflag.StringFlag{
			Name:   "packet-termination-time",
			Usage:  "The Packet Instance Termination Time",
			EnvVar: "PACKET_TERMINATION_TIME",
		},
		mcnflag.StringFlag{
			EnvVar: "PACKET_UA_PREFIX",
			Name:   "packet-ua-prefix",
			Usage:  fmt.Sprintf("Prefix the User-Agent in Packet API calls with some 'product/version'"),
		},
	}
}

func (d *Driver) DriverName() string {
	return "packet"
}

func (d *Driver) SetConfigFromFlags(flags drivers.DriverOptions) error {
	if strings.Contains(flags.String("packet-os"), "coreos") {
		d.SSHUser = "core"
	}
	if strings.Contains(flags.String("packet-os"), "rancher") {
		d.SSHUser = "rancher"
	}

	d.ApiKey = flags.String("packet-api-key")
	d.ProjectID = flags.String("packet-project-id")
	d.OperatingSystem = flags.String("packet-os")
	d.Facility = flags.String("packet-facility-code")
	d.BillingCycle = flags.String("packet-billing-cycle")
	d.UserAgentPrefix = flags.String("packet-ua-prefix")
	d.UserDataFile = flags.String("packet-userdata")

	d.Plan = flags.String("packet-plan")
	d.HardwareReserverationID = flags.String("packet-hw-reservation-id")

	d.SpotInstance = flags.Bool("packet-spot-instance")

	if d.SpotInstance {
		SpotPriceMax := flags.String("packet-spot-price-max")
		if SpotPriceMax == "" {
			d.SpotPriceMax = -1
		} else {
			SpotPriceMax, err := strconv.ParseFloat(SpotPriceMax, 64)
			if err != nil {
				return err
			}
			d.SpotPriceMax = SpotPriceMax
		}

		TerminationTime := flags.String("packet-termination-time")
		if TerminationTime == "" {
			d.TerminationTime = nil
		} else {
			layout := "2006-01-02T15:04:05.000Z"
			TerminationTime, err := time.Parse(layout, TerminationTime)
			if err != nil {
				return err
			}
			d.TerminationTime.Time = TerminationTime
		}
	}

	if d.ApiKey == "" {
		return fmt.Errorf("packet driver requires the --packet-api-key option")
	}
	if d.ProjectID == "" {
		return fmt.Errorf("packet driver requires the --packet-project-id option")
	}

	return nil
}

func (d *Driver) GetSSHHostname() (string, error) {
	return d.GetIP()
}

func (d *Driver) PreCreateCheck() error {
	if d.UserDataFile != "" {
		if _, err := os.Stat(d.UserDataFile); os.IsNotExist(err) {
			return fmt.Errorf("user-data file %s could not be found", d.UserDataFile)
		}
	}

	flavors, err := d.getOsFlavors()
	if err != nil {
		return err
	}
	if !stringInSlice(d.OperatingSystem, flavors) {
		return fmt.Errorf("specified --packet-os not one of %v", strings.Join(flavors, ", "))
	}

	if d.Facility == "any" {
		return nil
	}

	client := d.getClient()
	facilities, _, err := client.Facilities.List(nil)
	if err != nil {
		return err
	}
	for _, facility := range facilities {
		if facility.Code == d.Facility {
			return nil
		}
	}

	return fmt.Errorf("packet requires a valid facility")
}

func (d *Driver) Create() error {
	var userdata string
	if d.UserDataFile != "" {
		buf, err := ioutil.ReadFile(d.UserDataFile)
		if err != nil {
			return err
		}
		userdata = string(buf)
	}

	log.Info("Creating SSH key...")

	key, err := d.createSSHKey()
	if err != nil {
		return err
	}

	d.SSHKeyID = key.ID

	hardwareReservationId := ""
	//check if hardware reservation requested
	if d.HardwareReserverationID != "" {
		hardwareReservationId = d.HardwareReserverationID
	}

	client := d.getClient()
	createRequest := &packngo.DeviceCreateRequest{
		Hostname:              d.MachineName,
		Plan:                  d.Plan,
		HardwareReservationID: hardwareReservationId,
		Facility:              []string{d.Facility},
		OS:                    d.OperatingSystem,
		BillingCycle:          d.BillingCycle,
		ProjectID:             d.ProjectID,
		UserData:              userdata,
		Tags:                  d.Tags,
		SpotInstance:          d.SpotInstance,
		SpotPriceMax:          d.SpotPriceMax,
	}

	log.Info("Provisioning Packet server...")
	newDevice, _, err := client.Devices.Create(createRequest)
	if err != nil {
		//cleanup ssh keys if device faild
		if _, err := client.SSHKeys.Delete(d.SSHKeyID); err != nil {
			if er, ok := err.(*packngo.ErrorResponse); !ok || er.Response.StatusCode != 404 {
				return err
			}
		}
		return err
	}
	t0 := time.Now()

	d.DeviceID = newDevice.ID

	for {
		newDevice, _, err = client.Devices.Get(d.DeviceID, nil)
		if err != nil {
			return err
		}

		for _, ip := range newDevice.Network {
			if ip.Public && ip.AddressFamily == 4 {
				d.IPAddress = ip.Address
			}
		}

		if d.IPAddress != "" {
			break
		}

		time.Sleep(1 * time.Second)
	}

	log.Infof("Created device ID %s, IP address %s",
		newDevice.ID,
		d.IPAddress)

	log.Info("Waiting for Provisioning...")
	stage := float32(0)
	for {
		newDevice, _, err = client.Devices.Get(d.DeviceID, nil)
		if err != nil {
			return err
		}
		if newDevice.State == "provisioning" && stage != newDevice.ProvisionPer {
			stage = newDevice.ProvisionPer
			log.Debugf("Provisioning %v%% complete", newDevice.ProvisionPer)
		}
		if newDevice.State == "active" {
			log.Debugf("Device State: %s", newDevice.State)
			break
		}
		time.Sleep(10 * time.Second)
	}

	log.Debugf("Provision time: %v.", time.Since(t0))

	log.Debug("Waiting for SSH...")
	if err := drivers.WaitForSSH(d); err != nil {
		return err
	}

	return nil
}

func (d *Driver) createSSHKey() (*packngo.SSHKey, error) {
	sshKeyPath := d.GetSSHKeyPath()
	log.Debugf("Writing SSH Key to %s", sshKeyPath)

	if err := ssh.GenerateSSHKey(sshKeyPath); err != nil {
		return nil, err
	}

	publicKey, err := ioutil.ReadFile(sshKeyPath + ".pub")
	if err != nil {
		return nil, err
	}

	createRequest := &packngo.SSHKeyCreateRequest{
		Label: fmt.Sprintf("docker machine: %s", d.MachineName),
		Key:   string(publicKey),
	}

	key, _, err := d.getClient().SSHKeys.Create(createRequest)
	if err != nil {
		return key, err
	}

	return key, nil
}

func (d *Driver) GetURL() (string, error) {
	ip, err := d.GetIP()
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("tcp://%s:2376", ip), nil
}

func (d *Driver) GetIP() (string, error) {
	if d.IPAddress == "" {
		return "", fmt.Errorf("IP address is not set")
	}
	return d.IPAddress, nil
}

func (d *Driver) GetState() (state.State, error) {
	device, _, err := d.getClient().Devices.Get(d.DeviceID, nil)
	if err != nil {
		return state.Error, err
	}

	switch device.State {
	case "queued", "provisioning", "powering_on":
		return state.Starting, nil
	case "active":
		return state.Running, nil
	case "powering_off":
		return state.Stopping, nil
	case "inactive":
		return state.Stopped, nil
	}
	return state.None, nil
}

func (d *Driver) Start() error {
	_, err := d.getClient().Devices.PowerOn(d.DeviceID)
	return err
}

func (d *Driver) Stop() error {
	_, err := d.getClient().Devices.PowerOff(d.DeviceID)
	return err
}

func (d *Driver) Remove() error {
	client := d.getClient()

	if _, err := client.SSHKeys.Delete(d.SSHKeyID); err != nil {
		if er, ok := err.(*packngo.ErrorResponse); !ok || er.Response.StatusCode != 404 {
			return err
		}
	}

	if _, err := client.Devices.Delete(d.DeviceID); err != nil {
		if er, ok := err.(*packngo.ErrorResponse); !ok || er.Response.StatusCode != 404 {
			return err
		}
	}
	return nil
}

func (d *Driver) Restart() error {
	_, err := d.getClient().Devices.Reboot(d.DeviceID)
	return err
}

func (d *Driver) Kill() error {
	_, err := d.getClient().Devices.PowerOff(d.DeviceID)
	return err
}

func (d *Driver) GetDockerConfigDir() string {
	return dockerConfigDir
}

func (d *Driver) getClient() *packngo.Client {
	client := packngo.NewClientWithAuth(consumerToken, d.ApiKey, nil)
	userAgent := fmt.Sprintf("docker-machine-driver-%s/%s %s", d.DriverName(), version, client.UserAgent)

	if len(d.UserAgentPrefix) > 0 {
		userAgent = fmt.Sprintf("%s %s", d.UserAgentPrefix, userAgent)
	}

	client.UserAgent = userAgent
	return client
}

func (d *Driver) getOsFlavors() ([]string, error) {
	operatingSystems, _, err := d.getClient().OperatingSystems.List()
	if err != nil {
		return nil, err
	}

	supportedDistros := []string{
		"centos",
		"coreos",
		"debian",
		"opensuse",
		"rancher",
		"ubuntu",
	}
	flavors := make([]string, 0, len(operatingSystems))
	for _, flavor := range operatingSystems {
		if stringInSlice(flavor.Distro, supportedDistros) {
			flavors = append(flavors, flavor.Slug)
		}
	}
	return flavors, nil
}

func stringInSlice(a string, list []string) bool {
	for _, b := range list {
		if b == a {
			return true
		}
	}
	return false
}


pkg/drivers/metal/metal.go
// SPDX-License-Identifier: BSD-3-Clause

package metal

import (
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/docker/machine/libmachine/drivers"
	"github.com/docker/machine/libmachine/log"
	"github.com/docker/machine/libmachine/mcnflag"
	"github.com/docker/machine/libmachine/ssh"
	"github.com/docker/machine/libmachine/state"
	"github.com/packethost/packngo"
)

const (
	dockerConfigDir = "/etc/docker"
	consumerToken   = "24e70949af5ecd17fe8e867b335fc88e7de8bd4ad617c0403d8769a376ddea72"
)

var (
	// version is set by goreleaser at build time
	version = "devel"

	// build time check that the Driver type implements the Driver interface
	_ drivers.Driver = &Driver{}
)

type Driver struct {
	*drivers.BaseDriver
	ApiKey                  string
	ProjectID               string
	Plan                    string
	HardwareReserverationID string
	Facility                string
	OperatingSystem         string
	BillingCycle            string
	DeviceID                string
	UserData                string
	Tags                    []string
	CaCertPath              string
	SSHKeyID                string
	UserDataFile            string
	UserAgentPrefix         string
	SpotInstance            bool
	SpotPriceMax            float64
	TerminationTime         *packngo.Timestamp
}

// NewDriver is a backward compatible Driver factory method.  Using
// new(metal.Driver) is preferred.
func NewDriver(hostName, storePath string) *Driver {
	return &Driver{
		BaseDriver: &drivers.BaseDriver{
			MachineName: hostName,
			StorePath:   storePath,
		},
	}
}

func (d *Driver) GetCreateFlags() []mcnflag.Flag {
	return []mcnflag.Flag{
		mcnflag.StringFlag{
			Name:   "metal-api-key",
			Usage:  "Equinix Metal api key",
			EnvVar: "METAL_AUTH_TOKEN",
		},
		mcnflag.StringFlag{
			Name:   "metal-project-id",
			Usage:  "Equinix Metal Project Id",
			EnvVar: "METAL_PROJECT_ID",
		},
		mcnflag.StringFlag{
			Name:   "metal-os",
			Usage:  "Equinix Metal OS",
			Value:  "ubuntu_16_04",
			EnvVar: "METAL_OS",
		},
		mcnflag.StringFlag{
			Name:   "metal-facility-code",
			Usage:  "Equinix Metal facility code",
			Value:  "ewr1",
			EnvVar: "METAL_FACILITY_CODE",
		},
		mcnflag.StringFlag{
			Name:   "metal-plan",
			Usage:  "Equinix Metal Server Plan",
			Value:  "baremetal_0",
			EnvVar: "METAL_PLAN",
		},
		mcnflag.StringFlag{
			Name:   "metal-hw-reservation-id",
			Usage:  "Equinix Metal Reserved hardware ID",
			EnvVar: "METAL_HW_ID",
		},
		mcnflag.StringFlag{
			Name:   "metal-billing-cycle",
			Usage:  "Equinix Metal billing cycle, hourly or monthly",
			Value:  "hourly",
			EnvVar: "METAL_BILLING_CYCLE",
		},
		mcnflag.StringFlag{
			Name:   "metal-userdata",
			Usage:  "Path to file with cloud-init user-data",
			EnvVar: "METAL_USERDATA",
		},
		mcnflag.BoolFlag{
			Name:   "metal-spot-instance",
			Usage:  "Request a Equinix Metal Spot Instance",
			EnvVar: "METAL_SPOT_INSTANCE",
		},
		mcnflag.StringFlag{
			Name:   "metal-spot-price-max",
			Usage:  "The maximum Equinix Metal Spot Price",
			EnvVar: "METAL_SPOT_PRICE_MAX",
		},
		mcnflag.StringFlag{
			Name:   "metal-termination-time",
			Usage:  "The Equinix Metal Instance Termination Time",
			EnvVar: "METAL_TERMINATION_TIME",
		},
		mcnflag.StringFlag{
			EnvVar: "METAL_UA_PREFIX",
			Name:   "metal-ua-prefix",
			Usage:  "Prefix the User-Agent in Equinix Metal API calls with some 'product/version'",
		},
	}
}

func (d *Driver) DriverName() string {
	return "metal"
}

func (d *Driver) SetConfigFromFlags(flags drivers.DriverOptions) error {
	if strings.Contains(flags.String("metal-os"), "coreos") {
		d.SSHUser = "core"
	}
	if strings.Contains(flags.String("metal-os"), "rancher") {
		d.SSHUser = "rancher"
	}

	d.ApiKey = flags.String("metal-api-key")
	d.ProjectID = flags.String("metal-project-id")
	d.OperatingSystem = flags.String("metal-os")
	d.Facility = flags.String("metal-facility-code")
	d.BillingCycle = flags.String("metal-billing-cycle")
	d.UserAgentPrefix = flags.String("metal-ua-prefix")
	d.UserDataFile = flags.String("metal-userdata")

	d.Plan = flags.String("metal-plan")
	d.HardwareReserverationID = flags.String("metal-hw-reservation-id")

	d.SpotInstance = flags.Bool("metal-spot-instance")

	if d.SpotInstance {
		SpotPriceMax := flags.String("metal-spot-price-max")
		if SpotPriceMax == "" {
			d.SpotPriceMax = -1
		} else {
			SpotPriceMax, err := strconv.ParseFloat(SpotPriceMax, 64)
			if err != nil {
				return err
			}
			d.SpotPriceMax = SpotPriceMax
		}

		TerminationTime := flags.String("metal-termination-time")
		if TerminationTime == "" {
			d.TerminationTime = nil
		} else {
			layout := "2006-01-02T15:04:05.000Z"
			TerminationTime, err := time.Parse(layout, TerminationTime)
			if err != nil {
				return err
			}
			d.TerminationTime.Time = TerminationTime
		}
	}

	if d.ApiKey == "" {
		return fmt.Errorf("metal driver requires the --metal-api-key option")
	}
	if d.ProjectID == "" {
		return fmt.Errorf("metal driver requires the --metal-project-id option")
	}

	return nil
}

func (d *Driver) GetSSHHostname() (string, error) {
	return d.GetIP()
}

func (d *Driver) PreCreateCheck() error {
	if d.UserDataFile != "" {
		if _, err := os.Stat(d.UserDataFile); os.IsNotExist(err) {
			return fmt.Errorf("user-data file %s could not be found", d.UserDataFile)
		}
	}

	flavors, err := d.getOsFlavors()
	if err != nil {
		return err
	}
	if !stringInSlice(d.OperatingSystem, flavors) {
		return fmt.Errorf("specified --metal-os not one of %v", strings.Join(flavors, ", "))
	}

	if d.Facility == "any" {
		return nil
	}

	client := d.getClient()
	facilities, _, err := client.Facilities.List(nil)
	if err != nil {
		return err
	}
	for _, facility := range facilities {
		if facility.Code == d.Facility {
			return nil
		}
	}

	return fmt.Errorf("metal requires a valid facility")
}

func (d *Driver) Create() error {
	var userdata string
	if d.UserDataFile != "" {
		buf, err := ioutil.ReadFile(d.UserDataFile)
		if err != nil {
			return err
		}
		userdata = string(buf)
	}

	log.Info("Creating SSH key...")

	key, err := d.createSSHKey()
	if err != nil {
		return err
	}

	d.SSHKeyID = key.ID

	hardwareReservationId := ""
	//check if hardware reservation requested
	if d.HardwareReserverationID != "" {
		hardwareReservationId = d.HardwareReserverationID
	}

	client := d.getClient()
	createRequest := &packngo.DeviceCreateRequest{
		Hostname:              d.MachineName,
		Plan:                  d.Plan,
		HardwareReservationID: hardwareReservationId,
		Facility:              []string{d.Facility},
		OS:                    d.OperatingSystem,
		BillingCycle:          d.BillingCycle,
		ProjectID:             d.ProjectID,
		UserData:              userdata,
		Tags:                  d.Tags,
		SpotInstance:          d.SpotInstance,
		SpotPriceMax:          d.SpotPriceMax,
	}

	log.Info("Provisioning Equinix Metal server...")
	newDevice, _, err := client.Devices.Create(createRequest)
	if err != nil {
		//cleanup ssh keys if device faild
		if _, err := client.SSHKeys.Delete(d.SSHKeyID); err != nil {
			if er, ok := err.(*packngo.ErrorResponse); !ok || er.Response.StatusCode != 404 {
				return err
			}
		}
		return err
	}
	t0 := time.Now()

	d.DeviceID = newDevice.ID

	for {
		newDevice, _, err = client.Devices.Get(d.DeviceID, nil)
		if err != nil {
			return err
		}

		for _, ip := range newDevice.Network {
			if ip.Public && ip.AddressFamily == 4 {
				d.IPAddress = ip.Address
			}
		}

		if d.IPAddress != "" {
			break
		}

		time.Sleep(1 * time.Second)
	}

	log.Infof("Created device ID %s, IP address %s",
		newDevice.ID,
		d.IPAddress)

	log.Info("Waiting for Provisioning...")
	stage := float32(0)
	for {
		newDevice, _, err = client.Devices.Get(d.DeviceID, nil)
		if err != nil {
			return err
		}
		if newDevice.State == "provisioning" && stage != newDevice.ProvisionPer {
			stage = newDevice.ProvisionPer
			log.Debugf("Provisioning %v%% complete", newDevice.ProvisionPer)
		}
		if newDevice.State == "active" {
			log.Debugf("Device State: %s", newDevice.State)
			break
		}
		time.Sleep(10 * time.Second)
	}

	log.Debugf("Provision time: %v.", time.Since(t0))

	log.Debug("Waiting for SSH...")
	if err := drivers.WaitForSSH(d); err != nil {
		return err
	}

	return nil
}

func (d *Driver) createSSHKey() (*packngo.SSHKey, error) {
	sshKeyPath := d.GetSSHKeyPath()
	log.Debugf("Writing SSH Key to %s", sshKeyPath)

	if err := ssh.GenerateSSHKey(sshKeyPath); err != nil {
		return nil, err
	}

	publicKey, err := ioutil.ReadFile(sshKeyPath + ".pub")
	if err != nil {
		return nil, err
	}

	createRequest := &packngo.SSHKeyCreateRequest{
		Label: fmt.Sprintf("docker machine: %s", d.MachineName),
		Key:   string(publicKey),
	}

	key, _, err := d.getClient().SSHKeys.Create(createRequest)
	if err != nil {
		return key, err
	}

	return key, nil
}

func (d *Driver) GetURL() (string, error) {
	ip, err := d.GetIP()
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("tcp://%s:2376", ip), nil
}

func (d *Driver) GetIP() (string, error) {
	if d.IPAddress == "" {
		return "", fmt.Errorf("IP address is not set")
	}
	return d.IPAddress, nil
}

func (d *Driver) GetState() (state.State, error) {
	device, _, err := d.getClient().Devices.Get(d.DeviceID, nil)
	if err != nil {
		return state.Error, err
	}

	switch device.State {
	case "queued", "provisioning", "powering_on":
		return state.Starting, nil
	case "active":
		return state.Running, nil
	case "powering_off":
		return state.Stopping, nil
	case "inactive":
		return state.Stopped, nil
	}
	return state.None, nil
}

func (d *Driver) Start() error {
	_, err := d.getClient().Devices.PowerOn(d.DeviceID)
	return err
}

func (d *Driver) Stop() error {
	_, err := d.getClient().Devices.PowerOff(d.DeviceID)
	return err
}

func (d *Driver) Remove() error {
	client := d.getClient()

	if _, err := client.SSHKeys.Delete(d.SSHKeyID); err != nil {
		if er, ok := err.(*packngo.ErrorResponse); !ok || er.Response.StatusCode != 404 {
			return err
		}
	}

	if _, err := client.Devices.Delete(d.DeviceID, false); err != nil {
		if er, ok := err.(*packngo.ErrorResponse); !ok || er.Response.StatusCode != 404 {
			return err
		}
	}
	return nil
}

func (d *Driver) Restart() error {
	_, err := d.getClient().Devices.Reboot(d.DeviceID)
	return err
}

func (d *Driver) Kill() error {
	_, err := d.getClient().Devices.PowerOff(d.DeviceID)
	return err
}

func (d *Driver) GetDockerConfigDir() string {
	return dockerConfigDir
}

func (d *Driver) getClient() *packngo.Client {
	client := packngo.NewClientWithAuth(consumerToken, d.ApiKey, nil)
	userAgent := fmt.Sprintf("docker-machine-driver-%s/%s %s", d.DriverName(), version, client.UserAgent)

	if len(d.UserAgentPrefix) > 0 {
		userAgent = fmt.Sprintf("%s %s", d.UserAgentPrefix, userAgent)
	}

	client.UserAgent = userAgent
	return client
}

func (d *Driver) getOsFlavors() ([]string, error) {
	operatingSystems, _, err := d.getClient().OperatingSystems.List()
	if err != nil {
		return nil, err
	}

	supportedDistros := []string{
		"centos",
		"coreos",
		"debian",
		"opensuse",
		"rancher",
		"ubuntu",
	}
	flavors := make([]string, 0, len(operatingSystems))
	for _, flavor := range operatingSystems {
		if stringInSlice(flavor.Distro, supportedDistros) {
			flavors = append(flavors, flavor.Slug)
		}
	}
	return flavors, nil
}

func stringInSlice(a string, list []string) bool {
	for _, b := range list {
		if b == a {
			return true
		}
	}
	return false
}
